src/storage.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * Manages using `localStorage` and `sessionStorage`.
+ *
+ * Has a few advantages over the native API, including
+ * - failing gracefully if storage is not available
+ * - you can save and retrieve any JSONable data
+ *
+ * All methods are available for both `localStorage` and `sessionStorage`.
+ */
+
+class Storage {
+	private storage: globalThis.Storage;
+	private available: boolean | undefined;
+
+	constructor(storage: globalThis.Storage) {
+		this.storage = storage;
+	}
+
+	/**
+	 * Check whether storage type is available.
+	 */
+	isAvailable(): boolean {
+		if (this.available !== undefined) {
+			return this.available;
+		}
+
+		try {
+			// https://mathiasbynens.be/notes/localstorage-pattern
+			const uid = new Date().toString();
+			this.storage.setItem(uid, uid);
+
+			// ensure value we get is the one we set
+			const result = this.storage.getItem(uid) === uid;
+			this.storage.removeItem(uid);
+
+			// if we haven't failed by now, `result` is the last thing we need to check
+			this.available = result;
+		} catch (err) {
+			this.available = false;
+		}
+
+		return this.available;
+	}
+
+	/**
+	 * Retrieve a value from the storage type.
+	 *
+	 * @param key - the name of the value
+	 */
+	get(key: string): unknown {
```

Comment:
I wondered here if there was a json value type to *avoid the `unknown`* but the closest I could find was https://github.com/microsoft/TypeScript/issues/1897#issuecomment-338650717 which tries to define it. I'm not sure if that's better than `unknown` in this case?

Reply:
any of the suggestions here give me `circularly references itself` errors :(

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/json-schema/index.d.ts